Expand setup instructions (#28)
DFE-Digital/dfe-analytics: README.md

## Configuration

### 1. Get a BigQuery project setup and add initial owners

Ask in Slack on the `#twd_data_insights` channel for someone to help you
procure a BigQuery instance in the `digital.education.gov.uk` Google Cloud
Organisation.

Ask for your `@digital.education.gov.uk` Google account to be setup as an owner
via the IAM and Admin settings. Add other team members as necessary.

#### Set up billing

You also need to set up your BigQuery instance with paid billing. This is
because `dfe-analytics` uses streaming, and streaming isn't allowed in the free
tier:

> ```bash

Reviewer: Don't need the `bash` bit here, it adds incorrect highlighting.

accessDenied: Access Denied: BigQuery BigQuery: Streaming insert is not allowed
in the free tier
```

> 

Reviewer: Might be worth adding that typically the delivery manager can help arrange this, although we can't include the name of who to email here, we should have it in some other documentation somewhere that's not publicly accessible.

### 2. Create a dataset and table

You should create separate datasets for each environment (dev, qa, preprod, prod etc.).

1. Open your project's BigQuery instance
2. Go to the Analysis -> SQL Workspace section
3. Tap on the 3 dots next to the project name, "Create data set"
4. Name it something like `APPLICATIONNAME_events_ENVIRONMENT`, such as `applyforqts_events_production`, and set the location to `europe-west2 (London)`
5. Select your new dataset
6. Open a new query execution tab.

> 7. Paste and run [create-events-table.sql](https://github.com/DFE-Digital/dfe-analytics/create-events-table.sql) to create a blank events table for dfe-analytics to stream data into (editing as appropriate).

Reviewer: ```suggestion
7. Edit [create-events-table.sql](https://github.com/DFE-Digital/dfe-analytics/create-events-table.sql) to add your table name, and run it to create a blank events table for dfe-analytics to stream data into (editing as appropriate).
```

| Title             | **BigQuery Analyst Custom**                        |
| Description       | Assigned to accounts used by analysts and SQL developers. |
| ID                | `bigquery_analyst_custom`                          |
| Role launch stage | General Availability                               |
| + Add permissions | See below                                          |

<details>
<summary>Permissions for bigquery_analyst_custom</summary>
    bigquery.datasets.get
    bigquery.datasets.getIamPolicy
    bigquery.datasets.updateTag
    bigquery.jobs.create
    bigquery.jobs.get
    bigquery.jobs.list
    bigquery.jobs.listAll
    bigquery.models.export
    bigquery.models.getData
    bigquery.models.getMetadata
    bigquery.models.list
    bigquery.routines.get
    bigquery.routines.list
    bigquery.savedqueries.create
    bigquery.savedqueries.delete
    bigquery.savedqueries.get
    bigquery.savedqueries.list
    bigquery.savedqueries.update
    bigquery.tables.createSnapshot
    bigquery.tables.export
    bigquery.tables.get
    bigquery.tables.getData
    bigquery.tables.getIamPolicy
    bigquery.tables.list
    bigquery.tables.restoreSnapshot

>     resourcemanager.projects.get

Reviewer: These seems to render out as one long list inside this `<details>` block, I think you need to use HTML here instead of markup, so maybe wrap this list of permissions up in a `<pre>...</pre>` block?

Reviewer: ```suggestion
<pre>
bigquery.datasets.get
bigquery.datasets.getIamPolicy
bigquery.datasets.updateTag
bigquery.jobs.create
bigquery.jobs.get
bigquery.jobs.list
bigquery.jobs.listAll
bigquery.models.export
bigquery.models.getData
bigquery.models.getMetadata
bigquery.models.list
bigquery.routines.get
bigquery.routines.list
bigquery.savedqueries.create
bigquery.savedqueries.delete
bigquery.savedqueries.get
bigquery.savedqueries.list
bigquery.savedqueries.update
bigquery.tables.createSnapshot
bigquery.tables.export
bigquery.tables.get
bigquery.tables.getData
bigquery.tables.getIamPolicy
bigquery.tables.list
bigquery.tables.restoreSnapshot
resourcemanager.projects.get
</pre>
```

</details>

#### Developer

| Field             | Value                                    |
| ----------------- | ---------------------------------------- |
| Title             | **BigQuery Developer Custom**            |
| Description       | Assigned to accounts used by developers. |
| ID                | `bigquery_developer_custom`              |
| Role launch stage | General Availability                     |
| + Add permissions | See below                                |

<details>
<summary>Permissions for bigquery_developer_custom</summary>

>     bigquery.connections.create
>     bigquery.connections.delete
>     bigquery.connections.get
>     bigquery.connections.getIamPolicy
>     bigquery.connections.list
>     bigquery.connections.update
>     bigquery.connections.updateTag
>     bigquery.connections.use
>     bigquery.datasets.create
>     bigquery.datasets.delete
>     bigquery.datasets.get
>     bigquery.datasets.getIamPolicy
>     bigquery.datasets.update
>     bigquery.datasets.updateTag
>     bigquery.jobs.create
>     bigquery.jobs.delete
>     bigquery.jobs.get
>     bigquery.jobs.list
>     bigquery.jobs.listAll
>     bigquery.jobs.update
>     bigquery.models.create
>     bigquery.models.delete
>     bigquery.models.export
>     bigquery.models.getData
>     bigquery.models.getMetadata
>     bigquery.models.list
>     bigquery.models.updateData
>     bigquery.models.updateMetadata
>     bigquery.models.updateTag
>     bigquery.routines.create
>     bigquery.routines.delete
>     bigquery.routines.get
>     bigquery.routines.list
>     bigquery.routines.update
>     bigquery.routines.updateTag
>     bigquery.savedqueries.create
>     bigquery.savedqueries.delete
>     bigquery.savedqueries.get
>     bigquery.savedqueries.list
>     bigquery.savedqueries.update
>     bigquery.tables.create
>     bigquery.tables.createSnapshot
>     bigquery.tables.delete
>     bigquery.tables.deleteSnapshot
>     bigquery.tables.export
>     bigquery.tables.get
>     bigquery.tables.getData
>     bigquery.tables.getIamPolicy
>     bigquery.tables.list
>     bigquery.tables.restoreSnapshot
>     bigquery.tables.setCategory
>     bigquery.tables.update
>     bigquery.tables.updateData
>     bigquery.tables.updateTag
>     resourcemanager.projects.get

Reviewer: ```suggestion
<pre>
bigquery.connections.create
bigquery.connections.delete
bigquery.connections.get
bigquery.connections.getIamPolicy
bigquery.connections.list
bigquery.connections.update
bigquery.connections.updateTag
bigquery.connections.use
bigquery.datasets.create
bigquery.datasets.delete
bigquery.datasets.get
bigquery.datasets.getIamPolicy
bigquery.datasets.update
bigquery.datasets.updateTag
bigquery.jobs.create
bigquery.jobs.delete
bigquery.jobs.get
bigquery.jobs.list
bigquery.jobs.listAll
bigquery.jobs.update
bigquery.models.create
bigquery.models.delete
bigquery.models.export
bigquery.models.getData
bigquery.models.getMetadata
bigquery.models.list
bigquery.models.updateData
bigquery.models.updateMetadata
bigquery.models.updateTag
bigquery.routines.create
bigquery.routines.delete
bigquery.routines.get
bigquery.routines.list
bigquery.routines.update
bigquery.routines.updateTag
bigquery.savedqueries.create
bigquery.savedqueries.delete
bigquery.savedqueries.get
bigquery.savedqueries.list
bigquery.savedqueries.update
bigquery.tables.create
bigquery.tables.createSnapshot
bigquery.tables.delete
bigquery.tables.deleteSnapshot
bigquery.tables.export
bigquery.tables.get
bigquery.tables.getData
bigquery.tables.getIamPolicy
bigquery.tables.list
bigquery.tables.restoreSnapshot
bigquery.tables.setCategory
bigquery.tables.update
bigquery.tables.updateData
bigquery.tables.updateTag
resourcemanager.projects.get
</pre>
```

</details>

#### Appender

| Field             | Value                                                      |
| ----------------- | ---------------------------------------------------------- |
| Title             | **BigQuery Appender Custom**                               |
| Description       | Assigned to accounts used by appenders (apps and scripts). |
| ID                | `bigquery_appender_custom`                                 |
| Role launch stage | General Availability                                       |
| + Add permissions | See below                                                  |

<details>
<summary>Permissions for bigquery_appender_custom</summary>

>     bigquery.datasets.get
>     bigquery.tables.get
>     bigquery.tables.updateData

Reviewer: ```suggestion
<pre>
bigquery.datasets.get
bigquery.tables.get
bigquery.tables.updateData
</pre>
```
